OpenRA.Mods.Common/Traits/Conditions/ExternalCondition.cs

```diff
@@ -58,65 +76,80 @@ public bool CanGrantCondition(Actor self, object source)
-			if (Info.SourceCap > 0 && permanentTokens.GetOrAdd(source).Count >= Info.SourceCap)
-				return false;
+			if (Info.SourceCap > 0)
+			{
+				HashSet<int> permanentTokensForSource;
+				if (permanentTokens.TryGetValue(source, out permanentTokensForSource) && permanentTokensForSource.Count >= Info.SourceCap)
+					return false;
+			}
 
-			if (Info.TotalCap > 0 && permanentTokens.Values.SelectMany(t => t).Count() >= Info.TotalCap)
+			if (Info.TotalCap > 0 && permanentTokens.Values.Sum(t => t.Count) >= Info.TotalCap)
 				return false;
 
 			return true;
 		}
 
 		public int GrantCondition(Actor self, object source, int duration = 0)
 		{
-			if (conditionManager == null || source == null || !CanGrantCondition(self, source))
+			if (!CanGrantCondition(self, source))
 				return ConditionManager.InvalidConditionToken;
 
-			var token = conditionManager.GrantCondition(self, Info.Condition, duration);
-			var permanent = permanentTokens.GetOrAdd(source);
+			var token = conditionManager.GrantCondition(self, Info.Condition);
+			HashSet<int> permanent;
+			permanentTokens.TryGetValue(source, out permanent);
 
 			if (duration > 0)
 			{
-				var timed = timedTokens.GetOrAdd(source);
-
-				// Remove expired tokens
-				timed.RemoveWhere(t => t.Expires < self.World.WorldTick);
-
 				// Check level caps
 				if (Info.SourceCap > 0)
 				{
-					if (permanent.Count + timed.Count >= Info.SourceCap)
+					var timedCount = timedTokens.Count(t => t.Source == source);
+					if ((permanent != null ? permanent.Count + timedCount : timedCount) >= Info.SourceCap)
 					{
-						var expire = timed.MinByOrDefault(t => t.Expires);
-						if (expire != null)
+						// Get timed token from the same source with closest expiration.
+						var expireIndex = timedTokens.FindIndex(t => t.Source == source);
+						if (expireIndex >= 0)
 						{
-							timed.Remove(expire);
-							if (conditionManager.TokenValid(self, expire.Token))
-								conditionManager.RevokeCondition(self, expire.Token);
+							var expireToken = timedTokens[expireIndex].Token;
+							timedTokens.RemoveAt(expireIndex);
+							if (conditionManager.TokenValid(self, expireToken))
+								conditionManager.RevokeCondition(self, expireToken);
 						}
 					}
 				}
 
 				if (Info.TotalCap > 0)
 				{
-					var totalCount = permanentTokens.Values.SelectMany(t => t).Count() + timedTokens.Values.SelectMany(t => t).Count();
+					var totalCount = permanentTokens.Values.Sum(t => t.Count) + timedTokens.Count;
 					if (totalCount >= Info.TotalCap)
 					{
 						// Prefer tokens from the same source
-						var expire = timedTokens.SelectMany(t => t.Value.Select(tt => new Tuple<object, TimedToken>(t.Key, tt)))
-							.MinByOrDefault(t => t.Item2.Expires);
-						if (expire != null)
+						if (timedTokens.Count > 0)
 						{
-							if (conditionManager.TokenValid(self, expire.Item2.Token))
-								conditionManager.RevokeCondition(self, expire.Item2.Token);
+							var expire = timedTokens[0].Token;
```

Comment:
This doesn't `// Prefer tokens from the same source`, but the original code didn't seem to either.  It would be good if we could fix that here.

Reply:
Should be able to repeat the same `expireIndex` check from above with a fallback to index 0 if not found.